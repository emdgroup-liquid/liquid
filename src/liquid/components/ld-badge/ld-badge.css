:host,
.ld-badge {
  --ld-badge-size-md: 1.0625rem;
  --ld-badge-size-lg: 1rem;
  --ld-badge-icon-size-md: 0.5625rem;
  --ld-badge-icon-size-lg: 0.75rem;
  --ld-badge-bg-col: var(--ld-thm-primary);
  --ld-badge-gap-md: var(--ld-sp-6);
  --ld-badge-gap-lg: var(--ld-sp-8);
  --ld-badge-padding-block-md: 0.2125rem;
  --ld-badge-padding-block-lg: 0.25rem;
  --ld-badge-padding-inline-md: 0.5rem;
  --ld-badge-padding-inline-lg: 0.625rem;

  align-items: center;
  background-color: var(--ld-badge-bg-col);
  border-radius: var(--ld-br-full);
  box-sizing: border-box;
  color: var(--ld-col-wht);
  display: inline-flex;
  font: var(--ld-typo-body-xs);
  font-weight: 700;
  gap: var(--ld-badge-gap-md);
  justify-content: center;
  line-height: 1.1;
  padding-block: var(--ld-badge-padding-block-md);
  padding-inline: var(--ld-badge-padding-inline-md);
  position: relative;

  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  > ld-icon,
  > .ld-icon {
    --ld-icon-size-md: var(--ld-badge-icon-size-md);
    flex-shrink: 0;
    margin: 0 calc(-1 * var(--ld-sp-8));
  }

  &.ld-badge--with-text {
    &,
    :host(&) {
      ::slotted(ld-icon),
      ::slotted(.ld-icon),
      > ld-icon,
      > .ld-icon {
        margin: 0 auto;
      }
    }
  }
}

.ld-badge--lg,
:host(.ld-badge--lg) {
  font: var(--ld-typo-body-s);
  font-weight: 700;
  gap: var(--ld-badge-gap-lg);
  /*height: var(--ld-badge-size-lg);*/
  line-height: 1.1;
  padding-block: var(--ld-badge-padding-block-lg);
  padding-inline: var(--ld-badge-padding-inline-lg);

  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  > ld-icon,
  > .ld-icon {
    --ld-icon-size-md: var(--ld-badge-icon-size-lg);
    transform: none;
  }
}

.ld-badge--brand-color,
:host(.ld-badge--brand-color) {
  background-color: var(--ld-col-wht);
  color: var(--ld-thm-primary);
}
