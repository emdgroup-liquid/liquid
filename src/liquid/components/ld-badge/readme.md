---
eleventyNavigation:
  key: Badge
  parent: Components
layout: layout.njk
title: Badge
permalink: components/ld-badge/
---

<link rel="stylesheet" href="css_components/ld-badge.css">
<link rel="stylesheet" href="css_components/ld-icon.css">

# ld-badge

---

## Default

{% example %}
<ld-badge icon="checkmark"></ld-badge>

<!-- React component -->

<LdBadge icon="checkmark" />

<!-- CSS component -->

<span class="ld-badge">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
</span>
{% endexample %}

## With custom icon

{% example %}
<ld-badge>
  <svg slot="icon" viewBox="0 0 24 24" fill="none">
    <rect x="1.5" y="1.5" width="21" height="21" rx="4.5" stroke="currentColor" stroke-width="3"/>
    <circle cx="12" cy="12" r="4.5" stroke="currentColor" stroke-width="3"/>
  </svg>
</ld-badge>

<!-- React component -->

<LdBadge>
  <svg {...{ slot: 'icon' }} viewBox="0 0 24 24" fill="none">
    <rect x="1.5" y="1.5" width="21" height="21" rx="4.5" stroke="currentColor" stroke-width="3"/>
    <circle cx="12" cy="12" r="4.5" stroke="currentColor" stroke-width="3"/>
  </svg>
</LdBadge>

<!-- CSS component -->

<span class="ld-badge">
  <svg class="ld-icon" viewBox="0 0 24 24" fill="none">
    <rect x="1.5" y="1.5" width="21" height="21" rx="4.5" stroke="currentColor" stroke-width="3"/>
    <circle cx="12" cy="12" r="4.5" stroke="currentColor" stroke-width="3"/>
  </svg>
</span>
{% endexample %}

## Different sizes

{% example %}
<ld-badge icon="checkmark"></ld-badge>

<ld-badge icon="checkmark" size="lg"></ld-badge>

<!-- React component -->

<LdBadge icon="checkmark" />

<LdBadge icon="checkmark" size="lg" />

<!-- CSS component -->

<span class="ld-badge">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
</span>

<span class="ld-badge ld-badge--lg">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
</span>
{% endexample %}

## With text

{% example %}
<ld-badge>Badge</ld-badge>

<ld-badge size="lg">Badge</ld-badge>

<ld-badge icon="checkmark">Badge</ld-badge>

<ld-badge icon="checkmark" size="lg">Badge</ld-badge>

<!-- React component -->

<LdBadge>Badge</LdBadge>

<LdBadge size="lg">Badge</LdBadge>

<LdBadge icon="checkmark">Badge</LdBadge>

<LdBadge icon="checkmark" size="lg">Badge</LdBadge>

<!-- CSS component -->

<span class="ld-badge">Badge</span>

<span class="ld-badge ld-badge--lg">Badge</span>

<span class="ld-badge ld-badge--with-text">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
  Badge
</span>

<span class="ld-badge ld-badge--with-text ld-badge--lg">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
  Badge
</span>
{% endexample %}

## On brand color

{% example '{ "background": "brand", "hasBorder": false }' %}
<ld-badge brand-color icon="checkmark"></ld-badge>

<ld-badge brand-color>Badge</ld-badge>

<!-- React component -->

<LdBadge brandColor icon="checkmark" />

<LdBadge brand-color>Badge</LdBadge>

<!-- CSS component -->

<span class="ld-badge ld-badge--brand-color">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
</span>

<span class="ld-badge ld-badge--brand-color">Badge</span>
{% endexample %}

## Custom color

{% example %}
<style>
  .ld-badge--success {
    --ld-badge-bg-col: var(--ld-thm-success);
  }
</style>

<ld-badge class="ld-badge--success" icon="checkmark"></ld-badge>

<ld-badge class="ld-badge--success">Badge</ld-badge>

<!-- React component -->

<style>{`
  .ld-badge--success {
    --ld-badge-bg-col: var(--ld-thm-success);
  }
`}</style>

<LdBadge className="ld-badge--success" icon="checkmark" />

<LdBadge className="ld-badge--success">Badge</LdBadge>

<!-- CSS component -->

<style>
  .ld-badge--success {
    --ld-badge-bg-col: var(--ld-thm-success);
  }
</style>

<span class="ld-badge ld-badge--success">
  <svg class="ld-icon" viewBox="0 0 14 14" fill="none"><path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/></svg>
</span>

<span class="ld-badge ld-badge--success">Badge</span>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                                              | Type               | Default     |
| ------------ | ------------- | -------------------------------------------------------- | ------------------ | ----------- |
| `brandColor` | `brand-color` | Defines badge custom color                               | `boolean`          | `undefined` |
| `icon`       | `icon`        | Icon name.                                               | `string`           | `undefined` |
| `key`        | `key`         | for tracking the node's identity when working with lists | `string \| number` | `undefined` |
| `ref`        | `ref`         | reference to component                                   | `any`              | `undefined` |
| `size`       | `size`        | The size of the badge, translated in rem                 | `"lg"`             | `undefined` |


## Shadow Parts

| Part     | Description |
| -------- | ----------- |
| `"icon"` |             |


## Dependencies

### Depends on

- [ld-icon](../ld-icon)

### Graph
```mermaid
graph TD;
  ld-badge --> ld-icon
  style ld-badge fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
