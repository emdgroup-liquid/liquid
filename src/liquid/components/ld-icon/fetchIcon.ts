import { getAssetPath } from '@stencil/core'

const iconCache = {}
const requestCache = {}

export async function fetchIcon(icon: string): Promise<string> {
  if (iconCache[icon]) {
    return iconCache[icon]
  }
  if (!requestCache[icon]) {
    requestCache[icon] = fetch(getAssetPath(`./assets/${icon}.svg`))
      .then((resp) => resp.text())
      .catch((err) => {
        console.error(`"${icon}" is not a valid name`, err)
        return ''
      })
  }

  const path = await requestCache[icon]
  iconCache[icon] = path

  return path
}
