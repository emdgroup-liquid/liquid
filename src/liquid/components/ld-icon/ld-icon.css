:host {
  display: inline-flex;

  > div {
    display: inline-flex;
  }
}
