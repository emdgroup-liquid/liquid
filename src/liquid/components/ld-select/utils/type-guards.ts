export const isLdOption = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionElement | HTMLLdOptionInternalElement =>
  ["LD-OPTION", "LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroup = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupElement | HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP", "LD-OPTGROUP-INTERNAL"].includes(
    (el as HTMLElement)?.tagName,
  );

export const isLdOptionInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionInternalElement =>
  ["LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroupInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP-INTERNAL"].includes((el as HTMLElement)?.tagName);

type HTMLLdOptInternal =
  | HTMLLdOptionInternalElement
  | HTMLLdOptgroupInternalElement;
export const isLdOptInternalHidden = (
  opt: HTMLLdOptInternal,
): opt is
  | (HTMLLdOptInternal & {
      hidden: true;
    })
  | (HTMLLdOptInternal & {
      filtered: true;
    }) => {
  return opt.hidden || opt.filtered;
};
