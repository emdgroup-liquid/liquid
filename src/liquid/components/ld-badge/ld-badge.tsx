import { Component, Element, h, Host, Prop, State } from '@stencil/core'
import { HTMLStencilElement } from '@stencil/core/internal'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  assetsDirs: ['assets'],
  tag: 'ld-badge',
  styleUrl: 'ld-badge.css',
  shadow: true,
})
export class LdBadge {
  @Element() el: HTMLStencilElement

  private observer: MutationObserver

  /** Defines badge custom color */
  @Prop() brandColor?: boolean

  /** Icon name. */
  @Prop() icon?: string

  /** The size of the badge, translated in rem */
  @Prop() size?: 'lg'

  @State() hasCustomIcon = false
  @State() hasText = false

  private updateState = () => {
    this.hasCustomIcon = !!this.el.querySelector('[slot="icon"]')
    this.hasText = !!this.el.textContent.trim()
  }

  componentWillLoad() {
    this.observer = new MutationObserver(this.updateState)
    this.observer.observe(this.el, {
      subtree: true,
      childList: true,
      attributes: false,
    })

    this.updateState()
  }

  disconnectedCallback() {
    /* istanbul ignore if */
    if (this.observer) this.observer.disconnect()
  }

  render() {
    const cl = getClassNames([
      'ld-badge',
      this.size && `ld-badge--${this.size}`,
      this.hasText && 'ld-badge--with-text',
      this.brandColor && `ld-badge--brand-color`,
    ])

    return (
      <Host class={cl}>
        {(this.icon || this.hasCustomIcon) && (
          <ld-icon class="ld-badge__icon" part="icon" name={this.icon}>
            <slot name="icon"></slot>
          </ld-icon>
        )}
        <slot></slot>
      </Host>
    )
  }
}
