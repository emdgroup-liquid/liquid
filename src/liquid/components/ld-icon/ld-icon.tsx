import { Build, Component, Host, h, Prop, Watch, Element } from '@stencil/core'
import { getClassNames } from '../../utils/getClassNames'
import { fetchIcon } from './fetchIcon'

/**
 * @slot - (optional) Custom SVG icon (only valid without name prop).
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part icon - Actual SVG element
 */
@Component({
  assetsDirs: ['assets'],
  tag: 'ld-icon',
  styleUrl: 'ld-icon.css',
  shadow: true,
})
export class LdIcon {
  @Element() element: HTMLElement

  /** The icon name. */
  @Prop() name: string = null

  /** Size of the icon. */
  @Prop() size?: 'sm' | 'lg'

  @Watch('name')
  private async loadIconPathData(): Promise<void> {
    if ((!Build.isBrowser && !Build.isTesting) || !this.name) {
      return
    }

    const div = document.createElement('div')
    const iconString = await fetchIcon(this.name)

    div.innerHTML = iconString.replace(
      '<svg',
      '<svg class="ld-icon__svg" part="icon"'
    )
    Array.from(this.element.shadowRoot.children).forEach((child) => {
      if (child.tagName !== 'STYLE') {
        this.element.shadowRoot.removeChild(child)
      }
    })
    this.element.shadowRoot.appendChild(div.firstChild)
  }

  async componentWillLoad(): Promise<void> {
    this.loadIconPathData()
  }

  render() {
    return (
      <Host
        class={getClassNames(['ld-icon', this.size && `ld-icon--${this.size}`])}
        role="presentation"
      >
        {!this.name && <slot></slot>}
      </Host>
    )
  }
}
