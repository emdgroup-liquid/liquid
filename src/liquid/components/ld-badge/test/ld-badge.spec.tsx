import { newSpecPage } from '@stencil/core/testing'
import { LdIcon } from '../../ld-icon/ld-icon'
import { LdBadge } from '../ld-badge'
import { getTriggerableMutationObserver } from '../../../utils/mutationObserver'

describe('ld-badge', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge icon="checkmark"></ld-badge>',
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders with custom icon', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: `<ld-badge>
        <svg slot="icon" viewBox="0 0 14 14">
          <path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"></path>
        </svg>
      </ld-badge>`,
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders with size lg', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge icon="checkmark" size="lg"></ld-badge>',
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders with text', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge>Badge</ld-badge>',
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders on brand color', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge brand-color icon="checkmark"></ld-badge>',
    })
    expect(page.root).toMatchSnapshot()
  })
  it('updates after addition of an icon', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge>Badge</ld-badge>',
    })
    await page.waitForChanges()

    const icon = page.doc.createElement('ld-icon')
    icon.setAttribute('slot', 'icon')
    icon.setAttribute('name', 'placeholder')

    page.root.appendChild(icon)
    getTriggerableMutationObserver().trigger([])
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })
  it('updates after addition of text', async () => {
    const page = await newSpecPage({
      components: [LdBadge, LdIcon],
      html: '<ld-badge icon="checkmark"></ld-badge>',
    })
    await page.waitForChanges()

    const span = page.doc.createElement('span')
    span.innerHTML = 'Badge'

    page.root.appendChild(span)
    getTriggerableMutationObserver().trigger([])
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })
})
