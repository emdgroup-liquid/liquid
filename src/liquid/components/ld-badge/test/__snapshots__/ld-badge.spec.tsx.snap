// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-badge renders 1`] = `
<ld-badge class="ld-badge" icon="checkmark">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
</ld-badge>
`;

exports[`ld-badge renders on brand color 1`] = `
<ld-badge brand-color="" class="ld-badge ld-badge--brand-color" icon="checkmark">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
</ld-badge>
`;

exports[`ld-badge renders with custom icon 1`] = `
<ld-badge class="ld-badge">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        <slot></slot>
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
  <svg slot="icon" viewBox="0 0 14 14">
    <path d="m12 4-6.592 6L2 6.6396" stroke="currentcolor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
  </svg>
</ld-badge>
`;

exports[`ld-badge renders with size lg 1`] = `
<ld-badge class="ld-badge ld-badge--lg" icon="checkmark" size="lg">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
</ld-badge>
`;

exports[`ld-badge renders with text 1`] = `
<ld-badge class="ld-badge ld-badge--with-text">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  Badge
</ld-badge>
`;

exports[`ld-badge updates after addition of an icon 1`] = `
<ld-badge class="ld-badge ld-badge--with-text">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        <slot></slot>
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
  Badge
  <ld-icon class="ld-icon" name="placeholder" role="presentation" slot="icon">
    <mock:shadow-root>
      Not Found
    </mock:shadow-root>
  </ld-icon>
</ld-badge>
`;

exports[`ld-badge updates after addition of text 1`] = `
<ld-badge class="ld-badge ld-badge--with-text" icon="checkmark">
  <mock:shadow-root>
    <ld-icon class="ld-badge__icon ld-icon" part="icon" role="presentation">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
      <slot name="icon"></slot>
    </ld-icon>
    <slot></slot>
  </mock:shadow-root>
  <span>
    Badge
  </span>
</ld-badge>
`;
